pkg/controller/certificatesigningrequests/vault/vault.go

```diff
@@ -0,0 +1,159 @@
+	certificatesclient "k8s.io/client-go/kubernetes/typed/certificates/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/record"
+
+	apiutil "github.com/jetstack/cert-manager/pkg/api/util"
+	cmapi "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
+	experimentalapi "github.com/jetstack/cert-manager/pkg/apis/experimental/v1alpha1"
+	controllerpkg "github.com/jetstack/cert-manager/pkg/controller"
+	"github.com/jetstack/cert-manager/pkg/controller/certificatesigningrequests"
+	"github.com/jetstack/cert-manager/pkg/controller/certificatesigningrequests/util"
+	internalvault "github.com/jetstack/cert-manager/pkg/internal/vault"
+	logf "github.com/jetstack/cert-manager/pkg/logs"
+	"github.com/jetstack/cert-manager/pkg/util/pki"
+)
+
+const (
+	CSRControllerName = "certificatesigningrequests-issuer-vault"
+)
+
+type signingFn func(*x509.Certificate, *x509.Certificate, crypto.PublicKey, interface{}) ([]byte, *x509.Certificate, error)
+
+// Vault is a controller for signing Kubernetes CertificateSigningRequest
+// using Vault Issuers.
+type Vault struct {
+	issuerOptions controllerpkg.IssuerOptions
+	secretsLister corelisters.SecretLister
+
+	recorder record.EventRecorder
+
+	certClient    certificatesclient.CertificateSigningRequestInterface
+	clientBuilder internalvault.ClientBuilder
+}
+
+func init() {
+	controllerpkg.Register(CSRControllerName, func(ctx *controllerpkg.Context) (controllerpkg.Interface, error) {
+		return controllerpkg.NewBuilder(ctx, CSRControllerName).
+			For(certificatesigningrequests.New(apiutil.IssuerVault, NewVault(ctx))).
+			Complete()
+	})
+}
+
+func NewVault(ctx *controllerpkg.Context) *Vault {
+	return &Vault{
+		issuerOptions: ctx.IssuerOptions,
+		secretsLister: ctx.KubeSharedInformerFactory.Core().V1().Secrets().Lister(),
+		recorder:      ctx.Recorder,
+		certClient:    ctx.Client.CertificatesV1().CertificateSigningRequests(),
+		clientBuilder: internalvault.New,
+	}
+}
+
+// Sign attempts to sign the given CertificateSigningRequest based on the
+// provided Vault Issuer or ClusterIssuer. This function updates the
+// CertificateSigningRequest resource if signing was successful. Returns an
+// error which, if not nil, should trigger a retry.
+func (v *Vault) Sign(ctx context.Context, csr *certificatesv1.CertificateSigningRequest, issuerObj cmapi.GenericIssuer) error {
+	log := logf.FromContext(ctx, "sign")
+	log = logf.WithRelatedResource(log, issuerObj)
+
+	resourceNamespace := v.issuerOptions.ResourceNamespace(issuerObj)
+
+	client, err := v.clientBuilder(resourceNamespace, v.secretsLister, issuerObj)
+	if apierrors.IsNotFound(err) {
+		message := "Required secret resource not found"
+		log.Error(err, message)
+		v.recorder.Event(csr, corev1.EventTypeWarning, "SecretNotFound", message)
+		util.CertificateSigningRequestSetFailed(csr, "SecretNotFound", message)
+		_, err := v.certClient.UpdateStatus(ctx, csr, metav1.UpdateOptions{})
+		return err
```

Comment:
suggestion: this is a permanent error that we already handled, no need to return an error

```suggestion
        // No need to retry since this is a permanent error (until the user creates
        // the missing Secret).
		return nil
```

Reply:
We are returning the error from `v.certClient.UpdateStatus` which should be retried.